### ISSUE TITLE
Magneto 2.1.0 Indexing PHP "Segmentation fault (core dumped)" (Fix Me!)

### ISSUE THREAD

**Issue by acidreian666** [magento/magento2#5900]
Having a Large product database pushes the limits of Magento 2.
My full database is over 183,000 Products.
Importing 30,000 products at a time and Indexing issue become relevant.
Anything > then 30,000 products creates this error.
### Preconditions

1. Install LEMP stack **Ubuntu 16.04** (Nginx1.10.0 PHP7.08-FPM Mysql 5.7)
   **(Also Seg Fault Exists on Debian 8.5 with Nginx php7.0.9-FPM MariaDB-10.0)**
2. Setup Install Magento 2.1 via webinstaller
3. Set mode Developer
4. increase Memory limits and execution time for large product importing.( my config's found below)
5. sudo php -f bin/magento Static-content Deploy | chmod -R 777 .  | chown -R www-data:www-data . 
6. sudo php -f bin/magento cache:disable
7. sudo php -f bin/magento cache:flush
8. sudo php -f bin/magento indexer:reindex
### Steps to reproduce

1. Select & import 1-30k.csv ( import will take 5-10minutes and be successful then manually reindex will be successful)
2. Select & import 30-60k.csv ( import will take 5-10minutes and be successful then manually reindex will Fail at catalogsearch_fulltext)
3. Once "Segmentation fault (core dumped)" is reproduced run
   "sudo php -f bin/magento indexer:reset" to invalidate the index's to attempt another reindex.
### Expected result

1. Successful Reindex of catalogsearch_fulltext without a "Segmentation fault (core dumped)"
### Actual result

1. Actual Result is "Segmentation fault (core dumped)"
2. strace log via "strace sudo php -f bin/magento indexer:reindex catalogsearch_fulltext"
   [Seg fault.txt](https://github.com/magento/magento2/files/391627/Seg.fault.txt)
### Trouble Shooting that I have done.

This is only an issue using Magento 2.1.

Tested with
Magento 2.0.8 = No issues. tested with php7.08 & php5.6
Magento 2.1 = Seg fault with php7.08 & php5.6
### Updated

Magento 2.2.0-dev that i pulled 7-30-16 also have the Seg Fault issue when indexing.

![adb seg](https://cloud.githubusercontent.com/assets/16370331/17264876/12ae1912-55b8-11e6-99e3-953a2c08bfa5.png)
![seg fault](https://cloud.githubusercontent.com/assets/16370331/17264882/1da8cc18-55b8-11e6-96f7-5157641873fc.png)
### Below is the Config files used & 1-60,000 products csv files
1. Magento site config for nginx (remove .txt ext to use in sites available) 
   [magento.txt](https://github.com/magento/magento2/files/391509/magento.txt)
2. php cli php.ini (change .txt to .ini)
   [php.txt](https://github.com/magento/magento2/files/391513/php.txt)
3. nginx config nginx.conf( remove .txt extension)
   [nginx.conf.txt](https://github.com/magento/magento2/files/391514/nginx.conf.txt)
4. mysqld.cnf( remove .txt extension)
   [mysqld.cnf.txt](https://github.com/magento/magento2/files/391515/mysqld.cnf.txt)
5. 1-30k.csv just random generated Product data( remove .txt extension)
   [1-30k.csv.txt](https://github.com/magento/magento2/files/391517/1-30k.csv.txt)
6. 30-60k.csv continuation of generated data for more product data( remove .txt extension)
   [30-60k.csv.txt](https://github.com/magento/magento2/files/391518/30-60k.csv.txt)

**Comment by southerncomputer** [comment]
You should use https://github.com/Smile-SA/elasticsuite or EE elastic search method for that many products! Seriously stuffing that many products into mysql's full text table is murder on storage/speed!

How much ram do you have? I've had large category/catalogs eat up 2-3 MEGS of php ram - if you can afford it, raise your ram up to 3GB for php (both in nginx/php.ini) (also check your /etc/php-fpm.d/*.conf for any overrides on max memory.

Watch with TOP and keep an eye on peak memory used before it seg faults! It may be possible for the indexer to work in smaller chunks (search for batchsize in the code).

**Comment by acidreian666** [comment]
@southerncomputer Looking into ElasticSuite looks like it has some great advantages. 
But before trying to switch from my current lane.. This is Issue with the Magento 2.1 & 2.2 indexer I'm not over doing the amount of index-able products via Sql because on versions 2.0.2-2.0.8 of Magento have had no issues indexing via CLI that many products. Something has changed from the older versions Where i have in the past successfully indexed over 180k products.

I have set my NGINX & php to 24-30GB of ram and it does not stop a SEG Fault.
my current php.ini is set to 10GB of ram.

I'm requesting that some one actually fallow this topic and Reproduce this on there own temp Dev server. I can set this up using VMware in 10 minutes hopefully some one will test and be able to DEBUG this Seg fault further then my knowledge.

**Comment by acidreian666** [comment]
@andimov @magento-team  Have you or anyone else been able to reproduce this Seg Fault if so i would like to help contribute to a solution but with my knowledge I have exhausted skills. via Strace & GDB for debugging..

**Comment by tedxas** [comment]
Same issue here. Magento v2.1.1. Already had about 20,000 products in the database. Added another 90,000 with two CSV files, each with roughly 45,000 products each.

Attempts to reindex catalogsearch_fulltext results in "Segmentation fault (core dumped)". Now, no products show in the website.

Many thanks to any/all who can propose a solution!

**Comment by southerncomputer** [comment]
I get the same problem with EE 2.1.0 on manual re-index with 700K products, using Smile-SA Elasticsuite indexes with no problem not using 8gb of PHP ram like CE/EE 2.1.1 ! what gives?

**Comment by tedxas** [comment]
Resolution attempt: gave MySQL its own server instance and shut down MySQL on the webserver with hopes this would give PHP more "processing room" to complete the reindexing. Still results in Segmentation Fault, however.

**Comment by southerncomputer** [comment]
can you try this on command line:
php -d memory_limit=-1 bin/magento indexer:reindex catalogsearch_fulltext

then watch via top for the size of the php process when it dies? Mine died around 8GB but if you have a reduced memory size host or 32bit php you might see it die at 4GB!

**Comment by tedxas** [comment]
Workaround:  
- Went to System->Index Management, checked "Catalog Search" and set the Action dropdown to "Update on Save"
- Used the System->Import function to Delete the Products I had uploaded via two separate files
- Somewhere during all of that, Catalog Search indexed because my original Products appeared in the store again.
- Re-added the Products via the two CSV files mentioned earlier.
- Received a notification that the index needed re-indexing. Clicked on Products->Catalog and noticed the reindex notification had already disappeared.
- Went back to System->Index Management and noticed that everything had re-indexed already.

So, basically, reverting Catalog Search to Update on Save seemed to do the trick.  Hard to say, though, because I'm a Magento 2 newbie.

**Comment by choukalos** [comment]
Have you tried upping the temp table limit in MySQL?  I believe there's a similar bug that the teams are working where the work around is bumping that limit.  You might try bumping that to several GBs to see if that addresses the issue.  There's a bug being work that'll be back ported to 2.1.x that'll remove the need to bump temp table limit in MySQL.

+1 for elastic search.  +1 for Smile - they've done some cool search work.  We've done some preliminary performance investigations around search and found that when you get more then a few concurrent search requests MySQL starts degrading (Elastic much higher concurrent limit before that happens).  In addition elastic has minimal growth/consistent retrieval times over catalog size where as MySQL linearly increases significantly as catalog size grows.  Our goal with the mysql adapter is to support hobbyist/small merchants (up to several $M in annual revenue for a small/medium catalog); beyond that Elasticsearch (or other option) is the way to go.

**Comment by jaiswal-grv** [comment]
I am also getting the same error, I added 60000 products in my system and getting segmentation fault when reindexing using CLI any solutions...

**Comment by JstrzAtW2M** [comment]
Same issue with 40,000 products.

**Comment by rafeel** [comment]
magneto. epic

**Comment by elvinristi** [comment]
Duplicate issue of #5146

**Comment by idziakjakub** [comment]
I think you should try run reindex command NOT as root user or with sudo. I had the same problem ( with 55000 products) when I used root and when I changed a user to "www-data" reindexation worked fine.

**Comment by elvinristi** [comment]
Not an user issue. There is already patch released by Magento, not sure if included in 2.1.4.

**Comment by magento-engcom-team** [comment]
@acidreian666, thank you for your report.
The issue is already fixed in develop branch, 2.2.0, 2.1.9
